cidr-filtering-cni

```diff
@@ -0,0 +1,174 @@
+    done
+}
+
+# returns string in format <address>/<mask> for given CIDR range
+# parameters
+# $1 json string with cidr struct
+ip_from_json_to_str_with_mask () {
+    ip=$(echo $1 | jq -r .ip)
+    prefix=$(echo $1 | jq -r '."prefix-length"')
+    echo "${ip}/${prefix}"
+}
+
+# returns ip or ip6
+# parameters
+# $1 <string> ip address
+get_ip_version() {
+    if [[ $1 =~ .*:.* ]]
+    then 
+        echo "ip6"
+    else
+        echo "ip"
+    fi
+}
+
+create_table() {
+    echo "nft add table $1 $2"
+}
+
+create_base_chain() {
+    echo "nft add chain $1 $2 $3 { type filter hook $4 priority 0; policy accept; }"
+}
+
+create_chain() {
+    echo "nft add chain $1 $2 $3"
+}
+
+nft_add_rule() {
+    type="$1"
+    table="$2"
+    chain="$3"
+    set -- "${@:4}"
+    echo "nft add rule $type $table $chain $@"
+}
+
+create_ingress_rule() {
+    ip_address=$(ip_from_json_to_str_with_mask $1)
+    nft_cmd="nft add rule ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-ingress-from iifname $CNI_IFNAME $(get_ip_version ${ip_address}) saddr ${ip_address} counter meta mark set mark or 0x20000"
+    ip netns exec $CNI_CONTAINERID ${nft_cmd}
+
+    # TODO support ports properly, now just allow everything
+    nft_cmd="nft add rule ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-ingress-ports counter meta mark set meta mark | 0x00010000"
+    ip netns exec $CNI_CONTAINERID ${nft_cmd}
+}
+
+create_egress_rule() {
+    ip_address=$(ip_from_json_to_str_with_mask $1)
+    nft_cmd="nft add rule ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-egress-to oifname $CNI_IFNAME $(get_ip_version ${ip_address}) daddr ${ip_address} counter meta mark set meta mark | 0x00020000"
+    ip netns exec $CNI_CONTAINERID ${nft_cmd}
+
+    # TODO support ports properly, now just allow everything
+    nft_cmd="nft add rule ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-egress-ports counter meta mark set meta mark | 0x00010000"
+    ip netns exec $CNI_CONTAINERID ${nft_cmd}
+}
+
+case $CNI_COMMAND in
+ADD)
+    echo "CNI_NETNS: $CNI_NETNS" >> $tmpfile
+    echo "CNI_IFNAME: $CNI_IFNAME" >> $tmpfile
+    echo "CNI_CONTAINERID: $CNI_CONTAINERID" >> $tmpfile
+
+    mkdir -p /var/run/netns/
+    ln -sfT $CNI_NETNS /var/run/netns/$CNI_CONTAINERID
+
+    ip netns exec $CNI_CONTAINERID $(create_table ${NFT_TABLE_TYPE} ${NFT_TABLE})
+
+    ip netns exec $CNI_CONTAINERID $(create_base_chain ${NFT_TABLE_TYPE} ${NFT_TABLE} ${NFT_PREROUTING_CHAIN} ${NFT_INGRESS_HOOK})
+    ip netns exec $CNI_CONTAINERID $(create_base_chain ${NFT_TABLE_TYPE} ${NFT_TABLE} ${NFT_POSTROUTING_CHAIN} ${NFT_EGRESS_HOOK})
+
+    # ingress chains
+    ip netns exec $CNI_CONTAINERID $(create_chain ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-ingress)
+    ip netns exec $CNI_CONTAINERID $(create_chain ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-ingress)
+    ip netns exec $CNI_CONTAINERID $(create_chain ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-ingress-from)
+    ip netns exec $CNI_CONTAINERID $(create_chain ${NFT_TABLE_TYPE} ${NFT_TABLE} multi-0-ingress-ports)
```

Comment:
Ah thanks, I didn't realize that use case. This makes sense.

So in the end, the idea is to reference MultiNetworkPolicies in net-attach-def and the CNI applies the rules accordingly to the current state of the policies, does it sound right?

Reply:
That's the thing: no. 

`MultiNetworkPolicies` refer to net-attach-defs via a this [`policy-for`](https://github.com/k8snetworkplumbingwg/multi-networkpolicy#policy-for-annotation) annotation - so it is the other way around.

The intended implementation is - imo - clearly meant for a controller based approach. Using this API, you'd have to filter all the relevant policies that point to the NAD being used. It can be done, but it will take more time.

Lasty, one thing that I would also like to point out is I am unsure how much sense it makes to follow the multi-network-policy API at all - we're only allowing a subset of it (no namespace nor label based policies ...). 

The advantage of using the multi-network policy is quite obvious: users wanting to migrate to full-fledged solutions of it would be able to do so seamlessly. The disadvantage would be they could be confused / surprised at the fact only a subset of it is implemented.